### ISSUE TITLE
FR: CLI option to open new files in the existing instance of Neovide instead of creating a new one

### ISSUE THREAD

**Issue by chrisgrieser** [neovide/neovide#1586]
**Is your feature request related to a problem? Please describe.**
When Neovide is already open, and I use `neovide`  (the CLI) to open another file, another instance of Neovide is opened instead of opening neovide in a new buffer in the already existing instance.

This is very impractical, since I often accidentally end up with several neovide instances, which cannot be dealt with properly via the OS (all have the same icon and same app name for app switchers), and also I cannot switch between them with things like `:buffer`.

**Describe the solution you'd like**
a CLI option to open new files in the existing instance of Neovide instead of creating a new one

**Describe alternatives you've considered**
Only opening the *first* file via `neovide` and all other files from inside neovide? That is quite cumbersome to do though.

As [@bboles outlined in a different issue](https://github.com/neovide/neovide/issues/1562#issuecomment-1265910327), if the `Neovide.app` would use the "open with" feature on mac properly, one could open files with the app and it would open in the same instance. If that is going to be implemented, `open -a "Neovide" {filename}` could be a workaround for this.

**Comment by avalonv** [comment]
Breaking the current standard and messing up people's workflows is the easiest way to saturate this issues page with an angry mob. Instead of changing the default it would be better to expose an argument which does that, but I agree this would be a cool feature.

**Comment by tstm** [comment]
As it is,  Neovide does not seem to be able to run multiple windows in one instance, so doing this would definitely break expected behaviour?

**Comment by chrisgrieser** [comment]
> Breaking the current standard and messing up people's workflows is the easiest way to saturate this issues page with an angry mob. Instead of changing the default it would be better to expose an argument which does that, but otherwise I agree this would be a cool feature.

Yes, an option was actually what I meant. Edited for clarity.

**Comment by kemiller** [comment]
+1 To this.  I'd be OK with a special CLI command for this, a la VSCode or VimR, or a flag.  I'd expect it to open a new buffer, not a new macOS window.

**Comment by shemerey** [comment]
yep, would be grate feature to have

**Comment by 9mm** [comment]
> I'd expect it to open a new buffer, not a new macOS window.

I definitely disagree with this. If you want a new buffer you should just open a new buffer in the current window.

I actually need an entirely different window so it can go on other monitors, or so I can move projects around independently, but they should spawn out of the same icon so you dont get this:

<img width="677" alt="image" src="https://github.com/neovide/neovide/assets/1386207/1134c48c-a369-4e83-9326-93fa3179bddb">

I think the flag should enable functionality exactly like all the other Vim GUI's like MacVim, VimR, etc.

**Comment by 9mm** [comment]
As a sidenote, I just *HAD* to do this;

```
➜  ~ n; n; n; n;n;n; n; n; n; n;n;n; n; n; n; n;n;n; n; n; n; n;n;n; n; n; n; n;n;n; n; n; n; n;n;n; n; n; n; n;n;n; n; n; n; n;n;n; n; n; n; n;n;n; n; n; n; n;n;n; n; n; n; n;n;n; n; n; n; n;n;n; n; n; n; n;n;n; n; n; n; n;n;n; n; n; n; n;n;n;
```

When they open, they create a literal wave 🌊

<img width="1724" alt="image" src="https://github.com/neovide/neovide/assets/1386207/7a6608b7-9299-4ace-a936-136486808227">

**Comment by mambusskruj** [comment]
I think it would be nice to have an option that makes it possible to reuse (open) the existing neovide window if you call neovide to open in the CWD that already has opened inside other neovide windows.

**Comment by Delice0** [comment]
> I think it would be nice to have an option that makes it possible to reuse (open) the existing neovide window if you call neovide to open in the CWD that already was opened inside other neovide window.
> 
> This feature makes it possible to comfortably use Neovide as part of Tmux workflow with sessions and its windows. So you open Neovide window in CWD that you want and continue to use terminal session. Then switch to the existing Neovide window easily calling Neovide again.
> 
> For example, for me missing this kind of feature is a dealbreaker to use Neovide regularly.

Agree

**Comment by falcucci** [comment]
For the time being, the reason we are not implementing this is well explained [here](https://github.com/neovide/neovide/issues/2378#issuecomment-1965579707) and [here](https://github.com/neovide/neovide/issues/2378#issuecomment-1965576197).

Neohub is an available option to manage neovide instances, you can check it out [here](https://github.com/neovide/neovide/issues/2378#issuecomment-1963113733).

**Comment by chrisgrieser** [comment]
Due to the PR  #2395, the `Neovide.app` now can open files correctly, so `open -a "Neovide" {filename}` will avoid opening multiple windows. This solves the issue at least on macOS.

**Comment by chrisgrieser** [comment]
@fredizzimo sorry to bother you, but I do not think this issue is fully resolved. Just one case (opening via `open` on macOS) is solved, opening via `neovide` (the CLI) still spawns new windows

**Comment by fredizzimo** [comment]
Yes, re-opening this, I thought we had another cross-platform issue for it, since it's the same on all other platforms, we don't have a built-in way of re-using the same Neovide instance. We had this https://github.com/neovide/neovide/issues/1293, but it was closed in favour of this one.

With a few shell commands it's easy to support though:
1. Start the server
  `nvim --headless --listen localhost:9034&`
2. Start Neovide and connect
  `neovide --server localhost:9034`
3. Edit a file
  `nvim --server localhost:9034 --remote-tab $PWD/file.txt`
  or 
  `nvim --server localhost:9034 --remote-edit $PWD/file.txt`

You can replace the port 9034, by any free TCP port on your system, I just selected a number randomly. If you want multiple instances you can use multiple ports

Note, you need to full path to the file, if you started the instance in another folder, that's why I added `$PWD`

If you use it much, then a shell alias could be created
`alias neovide-open='nvim --server localhost:9034 --remote-tab' and used like this `neovide-open $PWD/file.txt`. But in practice you probably want something a bit smarter than that, that automatically converts relative paths to absolute, and supports multiple arguments. I leave that as an exercise for the reader.

**Comment by chrisgrieser** [comment]
Yeah, I am using sth similar to that as a workaround currently, though it is quite cumbersome. 

It's also not very beginner friendly for people starting out with nvim.

**Comment by lparry** [comment]
I've tried both
```bash
alias nvim='open -a "Neovide"'
```
and I've tried
```bash
function nvim() {
  local nvim_socket
  nvim_socket=/tmp/neovide.socket

  if [ ! -S $nvim_socket ]; then
    nvim --server $nvim_socket --remote-edit $PWD/file.txt
  else
    nvim --headless --listen $nvim_socket &
    neovide --server $nvim_socket
  fi
}
```
but both those options seem to have significantly worse performance for me than just running 
```bash
neovide $filename
```

Scrolling is buttery smooth via `neovide` but regularly chokes and stutters using either of the other options when navigating a huge ruby file (full of lines longer than lines have any right to be).

Would anyone have any suggestions why that might be, or what else I might try to get the normal neovide performance, while being able to add new buffers to the same instance from the cli?

**Comment by mooktakim** [comment]
Overall Neovide is the best GUI for neovim, but this one annoying thing makes it difficult to work with.
Its reasonable to expect to have multiple windows, and not have many icons on deck. You can't switch between windows with "cmd+`" (like every other app). There's also no way to know which window you are choosing.

Is there no way to resolve this issue? this is a very unusual behaviour for a macos app.

**Comment by fredizzimo** [comment]
The reason is mostly lack of developers on macOS. It's also technically a quite big task.

I'm not a macOS developer myself, but all the documentation I have seen suggest that in order to have multiple windows grouped together they need to be created by the same process. This is very different from the other operating systems, where the OS/Window Manager/Compositor can group the windows together in the same icon without any special application support.

This means that in order to support this, we need to re-structure the code so that we can maintain connections to several different Neovim instances, where each instance will render into its own window. It's not that hard to do, but we do use a few globals for example, which we would need to get rid of.

**Comment by mooktakim** [comment]
ok, thanks for the explanation @fredizzimo 👍

**Comment by falcucci** [comment]
Just to keep everyone in track of it, this is supposed to be working very soon in the multi-window refactor I am doing at [this](https://github.com/neovide/neovide/pull/2872) pull request. As a long as I am arriving at some stability of the core changes, the many Neovide icons (instances) visible in your dock will not be a problem anymore and this should work as only one instance for macOS.

Updates coming soon.

**Comment by kgfly** [comment]
Thank you @falcucci . Just wondering: your PR will also work on Linux/Windows, not just mac, correct?

**Comment by kgfly** [comment]
> yep, would be grate feature to have.
> 
> Solution in this case can be pretty trivial, since `neovide` relay on a nvim server any way we just need to define static listener and then send `--remote` command to it. In my case implementation was next. I've created script in `/usr/local/bin/v` and use my own socket.
> 
> #!/bin/sh
> 
> NVIM_SOCKET=/tmp/neovide.socket
> if [ ! -S $NVIM_SOCKET ]; then
>     neovide --noidle  -- --listen $NVIM_SOCKET $@
> else
>     nvim --server $NVIM_SOCKET --remote $@
> fi

I try Windows' corresponding things (AFAIK, windows does not support socket. It only supports pipe):

neovide --no-idle -- --listen "\\\\.\\pipe\\myneovim" 1.txt
nvim --server "\\\\.\\pipe\\myneovim" --remote 2.txt

But it does not work. Anyone has idea? Thanks.

**Comment by optimistiCli** [comment]
Meanwhile here's small wrapper script for opening files from command line on macOS in existing window (and tabless).

https://github.com/optimistiCli/nv

Hope someone will find it useful (-:

**Comment by kgfly** [comment]
Thanks. On Windows, I found these cmd work:

start /B nvim --headless --listen //./pipe/tmp/main-single-instance
neovide --server=tmp/main-single-instance
nvim --server "//./pipe/tmp/main-single-instance" --remote 1.txt
nvim --server "//./pipe/tmp/main-single-instance" --remote 2.txt
....

The only little annoying thing to me is: when closing, it prompts

![Image](https://github.com/user-attachments/assets/02516f3d-5e9b-4796-b3a9-4002def83c80)

How to set "Quit" as the default and disable the prompt?

(I am newbie to neovim :))

**Comment by fredizzimo** [comment]
@kgfly https://neovide.dev/configuration.html?highlight=Det#detach-on-quit

**Comment by Jimmy-Z** [comment]
I want to clarify that the original feature request was "open new files in existing instance AND window", while later it was hijacked to "open new files in existing instance but new windows" AKA multi-window support, I'd suggest separating this to a new issue to track both issues in a clearer way.

**Comment by falcucci** [comment]
>  while later it was hijacked to "open new files in existing instance but new windows" AKA multi-window support

you can be part of it! create an issue yourself to track it! we are just random people trying to make neovide great.

that said, I guess both features will be accomplished soon, that's something I am working on already.

**Comment by falcucci** [comment]
tonight this will be available on [**neovide nightly**](https://github.com/neovide/neovide/releases/tag/nightly) which we now can use `--reuse-instance` from the cli. 

also I am already working on smart project matching routing and new isolated windows.

note: tomorrow i'll travel to visit my family so the work will continue after on.

**Comment by varp** [comment]
I don't know how but the --chdir [functionality](https://neovide.dev/command-line-reference.html#working-directory) got broken on Nightly builds

**Comment by falcucci** [comment]
@varp the setting chdir isn't broken. for `--reuse-instance` we aren't sending the caller context like its correct `cwd`  which just doesnt resolve the new path correctly depending on your chdir value. this will be fixed soon.

edit: already available on [**nightly**](https://github.com/neovide/neovide/releases/tag/nightly)

**Comment by falcucci** [comment]
update: [you can](https://github.com/neovide/neovide/pull/3420) now create new windows as well from the cli also using `--reuse-instance --new-window` for separated projects you wish. i'll add some documentation later.

**Comment by varp** [comment]
What a nice day. Thank you so much

**Comment by varp** [comment]
> [@varp](https://github.com/varp) the setting chdir isn't broken. for `--reuse-instance` we aren't sending the caller context like its correct `cwd` which just doesnt resolve the new path correctly depending on your chdir value. this will be fixed soon.
> 
> edit: already available on [**nightly**](https://github.com/neovide/neovide/releases/tag/nightly)

Sorry for discussion on closed thread, but I've got the following wrapper:
```bash
nv() {
    if [[ "$_OS_NAME" != "Darwin" ]]; then
        echo "nv: wrapper only works on MacOS"
        exit 1
    else
        local dir=""
        if [[ ! -d $1 ]]; then 
            dir=$(dirname "$(realpath "$1")")
        else
            dir=$(realpath "$1")
        fi

        echo "$dir"

        open -n -a Neovide --env NEOVIDE_CHDIR="$dir" --args --no-tabs --reuse-instance --new-window --frame transparent --chdir "$dir" "${1}"
    fi
}
```

Neovide opens every new folder in a new window, but the directory doesn't switch correctly. The CWD is still set to the home folder path. I need to manually run `tcd <desired_directory>` to change the current directory for the current window. Maybe I am doing something wrong?

**Comment by falcucci** [comment]
> Maybe I am doing something wrong?

@varp you are doing nothing wrong regarding the chdir. the chdir worked fine for the old single window behavior, now i am refactoring that to the --new-window too, meaning that neovide was mixing those meanings, current cwd vs caller cwd which is now [fixed](https://github.com/neovide/neovide/pull/3422) in the handoff side.

you are going to have it per route instead of the only once started as before.

**Comment by ayevee** [comment]
@falcucci, do I understand correctly, that the feature has been implemented exclusively for macOS? Is there a chance to have it available for other operating systems as well?
